### PR TITLE
(RE-5816) Add a task for platform removal tickets

### PR DESCRIPTION
**Pull request by shermdog**
Added a rake task to facilitate the removal of supported platforms.

Usage:
```
PE_ONLY=true SERVER=true PLATFORM_TAG=remove_x86 JIRA_USER=user JIRA_INSTANCE=https://jira.host rake pl:platform_removal
```
PE_ONLY is for platforms that are only contained in PE
SERVER is for platforms that are also puppetserver / puppetdb hosts

### CODE REVIEW

**Comment by shermdog** [comment]
Some examples
```
17:48 $ PE_ONLY=false SERVER=false PLATFORM_TAG=remove-ignore-test-2 JIRA_USER=richard.sherman JIRA_INSTANCE=https://jira1-test.ops.puppetlabs.net rake pl:platform_removal
Logging in to https://jira1-test.ops.puppetlabs.net as richard.sherman
Password please: 
Okay trying to log in to https://jira1-test.ops.puppetlabs.net as richard.sherman ...
Creating new platform tickets based on:
{:pe_only=>false,
 :server=>false,
 :platform_tag=>"remove-ignore-test-2",
 :site=>"https://jira1-test.ops.puppetlabs.net",
 :username=>"richard.sherman",
 :base64_encoding=>"cmljaGFyZC5zaGVybWFuOlZ1bGdhclB5dGgwbiE="}
Main epic: CPR-321 (richard.sherman) - Remove remove-ignore-test-2
	CPR-322 () - Send email to community notifying of remove-ignore-test-2 removal
	PE-14484 () - Remove remove-ignore-test-2 from PE integration pipelines
	PA-245 () - Remove remove-ignore-test-2 from puppet-agent pipelines
	QENG-3531 () - Remove remove-ignore-test-2 from beaker-hostgenerator
	RE-6791 () - Remove remove-ignore-test-2 from puppetlabs-release
	RE-6792 () - Remove remove-ignore-test-2 platform definition from puppet-agent
	RE-6793 () - Remove remove-ignore-test-2 platform definition from pl-build-tools-vanagon
	RE-6794 () - Remove remove-ignore-test-2 platform definition from packaging
	RE-6795 () - Remove remove-ignore-test-2 from osmirror in puppetlabs-modules
	RE-6796 () - Remove remove-ignore-test-2 from vmpooler
	OPS-8366 () - Remove remove-ignore-test-2 from graphite (vmpooler)
	CPR-323 () - Remove remove-ignore-test-2 from repositories
	CPR-324 () - Send email to community confirming remove-ignore-test-2 removal
```
https://jira1-test.ops.puppetlabs.net/browse/CPR-321

```
17:04 $ PE_ONLY=false SERVER=false PLATFORM_TAG=remove-ignore-test-2 JIRA_USER=richard.sherman JIRA_INSTANCE=https://jira1-test.ops.puppetlabs.net rake pl:platform_removal
Logging in to https://jira1-test.ops.puppetlabs.net as richard.sherman
Password please: 
Okay trying to log in to https://jira1-test.ops.puppetlabs.net as richard.sherman ...
Creating new platform tickets based on:
{:pe_only=>"false",
 :server=>"false",
 :platform_tag=>"remove-ignore-test-2",
 :site=>"https://jira1-test.ops.puppetlabs.net",
 :username=>"richard.sherman",
 :base64_encoding=>"redacted"}
Main epic: RE-6784 (richard.sherman) - Remove remove-ignore-test-2
	CPR-318 () - Send email to community notifying of remove-ignore-test-2 removal
	PE-14483 () - Remove remove-ignore-test-2 from PE integration pipelines
	PA-244 () - Remove remove-ignore-test-2 from puppet-agent pipelines
	QENG-3530 () - Remove remove-ignore-test-2 from beaker-hostgenerator
	RE-6785 () - Remove remove-ignore-test-2 from puppetlabs-release
	RE-6786 () - Remove remove-ignore-test-2 platform definition from puppet-agent
	RE-6787 () - Remove remove-ignore-test-2 platform definition from pl-build-tools-vanagon
	RE-6788 () - Remove remove-ignore-test-2 platform definition from packaging
	RE-6789 () - Remove remove-ignore-test-2 from osmirror in puppetlabs-modules
	RE-6790 () - Remove remove-ignore-test-2 from vmpooler
	OPS-8365 () - Remove remove-ignore-test-2 from graphite (vmpooler)
	CPR-319 () - Remove remove-ignore-test-2 from repositories
	CPR-320 () - Send email to community confirming remove-ignore-test-2 removal
	SERVER-1196 () - Remove remove-ignore-test-2 from puppetserver pipelines
	PDB-2496 () - Remove remove-ignore-test-2 from puppetdb pipelines
	EZ-67 () - Remove remove-ignore-test-2 from ezbake build targets
```
https://jira1-test.ops.puppetlabs.net/browse/RE-6784

File: tasks/platform-removal.rake [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        blocks << subticket_hash[linked_ticket]['key']
+      end
+    end
+
+    # If this ticket already has an epic parent, we don't want to link it against the main epic
+    # since the epic parent *should* already be linked against the main ticket
+    if subticket[:epic_parent]
+      subticket[:epic_parent] = subticket_hash[subticket[:epic_parent]]['key']
+    else # otherwise, we definitely need to link it to the main epic ticket somehow
+      if subticket[:project] == 'RE' && (!subticket[:type] || subticket[:type] != 'Epic')
```
Comment (melissa) [inline tasks/platform-removal.rake]: you should also add the `CPR` project to this list

**Comment by shermdog** [comment]
Updated examples:

```
11:57 $ PE_ONLY=true SERVER=true PLATFORM_TAG=remove-ignore-test-6 EOL_DATE=2016-01-01 EOL_LINK=http://something.com/article JIRA_USER=richard.sherman JIRA_INSTANCE=https://jira1-test.ops.puppetlabs.net rake pl:platform_removal
Logging in to https://jira1-test.ops.puppetlabs.net as richard.sherman
Password please: 
Okay trying to log in to https://jira1-test.ops.puppetlabs.net as richard.sherman ...
Creating new platform tickets based on:
{:pe_only=>true,
 :server=>true,
 :platform_tag=>"remove-ignore-test-6",
 :eol_date=>"2016-01-01",
 :eol_link=>"http://something.com/article",
 :site=>"https://jira1-test.ops.puppetlabs.net",
 :username=>"richard.sherman",
 :base64_encoding=>"<redacted>="}
Main epic: RE-7002 (richard.sherman) - Remove remove-ignore-test-6
	CPR-333 () - Send email to community notifying of remove-ignore-test-6 removal
	PE-14921 () - Remove remove-ignore-test-6 from PE integration pipelines
	PA-279 () - Remove remove-ignore-test-6 from puppet-agent pipelines
	QENG-3633 () - Remove remove-ignore-test-6 from beaker-hostgenerator
	RE-7003 () - Remove remove-ignore-test-6 from puppetlabs-release
	RE-7004 () - Remove remove-ignore-test-6 platform definition from puppet-agent
	RE-7005 () - Remove remove-ignore-test-6 platform definition from pl-build-tools-vanagon
	RE-7006 () - Remove remove-ignore-test-6 platform definition from packaging
	RE-7007 () - Remove remove-ignore-test-6 from osmirror in puppetlabs-modules
	RE-7008 () - Remove remove-ignore-test-6 from vmpooler
	OPS-8641 () - Remove remove-ignore-test-6 from graphite (vmpooler)
	CPR-334 () - Remove remove-ignore-test-6 from repositories
	CPR-335 () - Send email to community confirming remove-ignore-test-6 removal
	SERVER-1252 () - Remove remove-ignore-test-6 from puppetserver pipelines
	PDB-2587 () - Remove remove-ignore-test-6 from puppetdb pipelines
	EZ-76 () - Remove remove-ignore-test-6 from ezbake build targets
```

https://jira1-test.ops.puppetlabs.net/browse/RE-7002

**Comment by melissa** [comment]
@shermdog :+1: but this needs a rebase

**Comment by shermdog** [comment]
Trying the fancy rebase button now :)

File: tasks/platform-removal.rake [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+  ]
+
+  # The subtickets for additonal server tasks
+  if vars[:server]
+    subtickets.push(
+      {
+        :short_name   => 'server_pipeline',
+        :project      => 'SERVER',
+        :summary      => "Remove #{vars[:platform_tag]} from puppetserver pipelines",
+        :description  => description[:server_pipeline],
```
Comment (melissa) [inline tasks/platform-removal.rake]: Oh, actually, these should be blocked by the initial email as well

File: tasks/platform-removal.rake [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+        :short_name   => 'pdb_pipeline',
+        :project      => 'PDB',
+        :summary      => "Remove #{vars[:platform_tag]} from puppetdb pipelines",
+        :description  => description[:pdb_pipeline],
+      },
+      {
+        :short_name   => 'ez',
+        :project      => 'EZ',
+        :summary      => "Remove #{vars[:platform_tag]} from ezbake build targets",
+        :description  => description[:ez],
```
Comment (melissa) [inline tasks/platform-removal.rake]: And this should be blocked by the pdb and puppetserver pipelines

File: tasks/platform-removal.rake [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+      :short_name   => 'initial_email',
+      :project      => 'CPR',
+      :summary      => "Send email to community notifying of #{vars[:platform_tag]} removal",
+      :description  => description[:initial_email],
+      :story_points => '1',
+    },
+    {
+      :short_name   => 'pe_pipeline',
+      :project      => 'PE',
+      :summary      => "Remove #{vars[:platform_tag]} from PE integration pipelines",
```
Comment (melissa) [inline tasks/platform-removal.rake]: This is just removing the agent package from the pe pipelines?

File: tasks/platform-removal.rake (hunk 2 above)
Comment (melissa) [inline tasks/platform-removal.rake]: I think you can also have this relationship define blockee tickets. So, these tickets should be blocking the beaker work

File: tasks/platform-removal.rake (hunk 4 above)
Comment (shermdog) [inline tasks/platform-removal.rake]: I believe that is correct.  That ticket was added to the list by Justin Stoller.

File: tasks/platform-removal.rake [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+end
+
+def create_platform_tickets(jira, vars)
+  description = {}
+  description[:initial_email] = <<-DOC
+Email should be sent out to puppet-users, puppet-dev, and puppet-announce notifying users the platform will no longer be supported.
+DOC
+
+  description[:pe_pipeline] = <<-DOC
+Platform needs to be removed from PE pipelines.
```
Comment (melissa) [inline tasks/platform-removal.rake]: I would prefer more description here, but it's not blocking this PR from being merged

**Comment by melissa** [comment]
cool, definitely :+1:

File: tasks/platform-removal.rake (hunk 5 above)
Comment (shermdog) [inline tasks/platform-removal.rake]: @justinstoller do you have any opinions on a more descriptive note for "Remove Platform from PE Pipelines"?

Comment (justinstoller) [inline tasks/platform-removal.rake]: I don't have any opinions.

In more detail, I think the work is:

If an agent platform, remove from integration pipelines including agent installation, frictionless install, and upgrade pipelines (all of those may have "primary" or "secondary" pipelines depending on the matrix configuration). I don't know if the upcoming PEZ work will need to be tweaked?

If a server platform, then the agent workflow should have already been ticketed and in addition pe-* pipelines will need to be updated (many of the projects on jenkins-enterprise and some others, notably: pe-puppet-server-extensions, pe-puppetdb-extensions, pe-console-services, and pe-orchestration-services). There may also be a set of "client tools" repos that need to be updated, don't know if we've split out that workflow from the server updates yet.

Basically, I think the person that works the ticket should start with the integration team and then work out from there the projects that are touching that platform and get "all" the applicable projects. Especially since the cursory examples of pipelines and projects I gave above will change from quarter to quarter.

File: tasks/platform-removal.rake [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+        :blocked_by   => ['initial_email'],
+        :blocks   => ['pa_pipeline'],
+      },
+      {
+        :short_name   => 'pdb_pipeline',
+        :project      => 'PDB',
+        :summary      => "Remove #{vars[:platform_tag]} from puppetdb pipelines",
+        :description  => description[:pdb_pipeline],
+        :blocked_by   => ['initial_email'],
+        :blocks   => ['pa_pipeline'],
```
Comment (melissa) [inline tasks/platform-removal.rake]: `:scrum_team => 'Release Engineering'` for the pdb, ez, puppetserver, and PA tickets, but not the qeng or PE tickets

**Comment by melissa** [comment]
:+1: pending travis
